# Objective
To automatically categorize/tag TED Talk videos based on a series of features provided in a sample dataset, link below.
Once the text data has been cleaned and prepared, various Classification models will be trained on the sample dataset.

There are two datasets provided: the main summary and the transcripts

**TED Talks: Summary**
https://www.kaggle.com/rounakbanik/ted-talks?select=ted_main.csv

**TED Talks: Transcripts**
https://www.kaggle.com/rounakbanik/ted-talks?select=transcripts.csv

# Problem Statement

The data provided, in its raw format, does not make for easy classification and statistical analysis of the various TED Talks. An automated, and simplified, approach that could quickly classify (tag) a given TED talk, based on other similar feature patterns, would allow for improved viewer experiences from accurate user preference alignment.

### Data Provided - Summary:

- **Comments:** The number of first level comments made on the talk
- **Description:** A blurb of what the talk is about
- **Duration:** The duration of the talk in seconds
- **Event:** The TED/TEDx event where the talk took place
- **Film_date:** The Unix timestamp of the filming
- **Languages:** The number of languages in which the talk is available
- **Main_speaker:** The first named speaker of the talk
- **Name:** The official name of the TED Talk; includes the title and the speaker
- **Num_speaker:** The number of speakers in the talk
- **Published_date:** The Unix timestamp for the publication of the talk on TED.com
- **Ratings:** Groups of ratings assinged to each talk
- **Related_talks:** Similar talks/links
- **Speaker_occupation:** Primary speaker occupation/focus
- **Tags:** Tags assinged to the talk
- **Title:** Title of TED talk
- **Url:** The URL of the talk
- **Views:** Total views for talk

### Data Provided - Transcript:
- **Transcript:** The official English transcript of the talk
- **URL:** The URL of the talk

In [92]:
# Import neccessary libaries

import pandas as pd # Dataframes
import numpy as np # Arrays and Math

# Pandas Profiling Summary report
from pandas_profiling import profile_report

# Visual libraries
import seaborn as sns
import plotly.graph_objects as go
import plotly.express as pe
import matplotlib.pyplot as plt

# set plt inline so plots will automatically run without the .show() command
%matplotlib inline 

# Warnings set to Ignore messages
import warnings
warnings.filterwarnings('ignore')

# Natural Language Processing (NLP)
import string 
import re # RegEX library for specific text isolation

from nltk.tokenize import sent_tokenize, word_tokenize # Tokenize sentences, then words with NLTK
from nltk.corpus import stopwords # Remove stopwords (the, a, is, are, etc.)
from nltk.stem import PorterStemmer # Reduce words back to original base word stem
from nltk import FreqDist # Counts frequency of word distributions

import textstat # Library to count to calculate statistics from text, readability, complexity, etc.

import ast # Abstract Syntax Tree - to help identify and segment information (text) imbedded in lists/dictionaries, etc.

#### Importing Dataset 1: Summary Details

In [93]:
# Import the TED summary dataset as a DataFrame
data1 = pd.read_csv('ted_main.csv')

#### Importing Dataset 2: Transcripts

In [94]:
# Import the TED transcripts dataset as a DataFrame
data2 = pd.read_csv('transcripts.csv')

In [95]:
# Create a copy DataFrame for each dataset, keeping data as backup if needed
TED_summ = data1.copy()

TED_trans = data2.copy()

In [96]:
# Set the dataframe summaries to include all columns - no truncating
pd.set_option('display.max_columns', None)

# View the top 5 rows of data - Summary
TED_summ.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

In [97]:
# View the top 5 rows of data - Transcripts
TED_trans.head()

transcript  \
0  Good morning. How are you?(Laughter)It's been ...   
1  Thank you so much, Chris. And it's truly a gre...   
2  (Music: "The Sound of Silence," Simon & Garfun...   
3  If you're here today — and I'm very happy that...   
4  About 10 years ago, I took on the task to teac...   

                                                 url  
0  https://www.ted.com/talks/ken_robinson_says_sc...  
1  https://www.ted.com/talks/al_gore_on_averting_...  
2  https://www.ted.com/talks/david_pogue_says_sim...  
3  https://www.ted.com/talks/majora_carter_s_tale...  
4  https://www.ted.com/talks/hans_rosling_shows_t...

In [98]:
# View the bottom 5 rows of data - Summary
TED_summ.tail()

comments                                        description  duration  \
2545        17  Between 2008 and 2016, the United States depor...       476   
2546         6  How can you study Mars without a spaceship? He...       290   
2547        10  Science fiction visions of the future show us ...       651   
2548        32  In an unmissable talk about race and politics ...      1100   
2549         8  With more than half of the world population li...       519   

             event   film_date  languages         main_speaker  \
2545       TED2017  1496707200          4     Duarte Geraldino   
2546       TED2017  1492992000          3  Armando Azua-Bustos   
2547       TED2017  1492992000          1       Radhika Nagpal   
2548  TEDxMileHigh  1499472000          1     Theo E.J. Wilson   
2549       TED2017  1492992000          1    Karoliina Korppoo   

                                                   name  num_speaker  \
2545  Duarte Geraldino: What we're missing in the de...            1   
2546  Armando Azua-Bustos: The most Martian place on...            1   
2547  Radhika Nagpal: What intelligent machines can ...            1   
2548  Theo E.J. Wilson: A black man goes undercover ...            1   
2549  Karoliina Korppoo: How a video game might help...            1   

      published_date                                            ratings  \
2545      1505851216  [{'id': 3, 'name': 'Courageous', 'count': 24},...   
2546      1505919737  [{'id': 22, 'name': 'Fascinating', 'count': 32...   
2547      1506006095  [{'id': 1, 'name': 'Beautiful', 'count': 14}, ...   
2548      1506024042  [{'id': 11, 'name': 'Longwinded', 'count': 3},...   
2549      1506092422  [{'id': 21, 'name': 'Unconvincing', 'count': 2...   

                                          related_talks   speaker_occupation  \
2545  [{'id': 2596, 'hero': 'https://pe.tedcdn.com/i...           Journalist   
2546  [{'id': 2491, 'hero': 'https://pe.tedcdn.com/i...       Astrobiologist   
2547  [{'id': 2346, 'hero': 'https://pe.tedcdn.com/i...    Robotics engineer   
2548  [{'id': 2512, 'hero': 'https://pe.tedcdn.com/i...  Public intellectual   
2549  [{'id': 2682, 'hero': 'https://pe.tedcdn.com/i...        Game designer   

                                                   tags  \
2545  ['TED Residency', 'United States', 'community'...   
2546  ['Mars', 'South America', 'TED Fellows', 'astr...   
2547  ['AI', 'ants', 'fish', 'future', 'innovation',...   
2548  ['Internet', 'TEDx', 'United States', 'communi...   
2549  ['cities', 'design', 'future', 'infrastructure...   

                                                  title  \
2545  What we're missing in the debate about immigra...   
2546                    The most Martian place on Earth   
2547  What intelligent machines can learn from a sch...   
2548       A black man goes undercover in the alt-right   
2549  How a video game might help us build better ci...   

                                                    url   views  
2545  https://www.ted.com/talks/duarte_geraldino_wha...  450430  
2546  https://www.ted.com/talks/armando_azua_bustos_...  417470  
2547  https://www.ted.com/talks/radhika_nagpal_what_...  375647  
2548  https://www.ted.com/talks/theo_e_j_wilson_a_bl...  419309  
2549  https://www.ted.com/talks/karoliina_korppoo_ho...  391721

In [99]:
# View the bottom 5 rows of data - Summary
TED_trans.tail()

transcript  \
2462  So, Ma was trying to explain something to me a...   
2463  This is a picture of a sunset on Mars taken by...   
2464  In my early days as a graduate student, I went...   
2465  I took a cell phone and accidentally made myse...   
2466  We humans are becoming an urban species, so ci...   

                                                    url  
2462  https://www.ted.com/talks/duarte_geraldino_wha...  
2463  https://www.ted.com/talks/armando_azua_bustos_...  
2464  https://www.ted.com/talks/radhika_nagpal_what_...  
2465  https://www.ted.com/talks/theo_e_j_wilson_a_bl...  
2466  https://www.ted.com/talks/karoliina_korppoo_ho...

In [100]:
# Inspect the overall shape of the data (rows/columns)
print("DataFrame Shapes:\n")
print("Summary Shape: {}".format(TED_summ.shape))
print("Transcripts Shape: {}".format(TED_trans.shape))

DataFrame Shapes:

Summary Shape: (2550, 17)
Transcripts Shape: (2467, 2)


- The summary file has 2,550 rows of data and 17 columns
- The Transcripts file has slightly less rows (2,467) and only 2 columns

**Both DataFrames have a URL column**, which can likely serve as the same column for merging both datasets into one overall DataFrame.  **83 rows of data will be lost** due to the Transcripts DataFrame having a lower overall row count.

In [101]:
# Use Pandas merge function to combine both datasets on the shared column, URL
TED_combined = pd.merge(TED_summ, TED_trans, how = 'inner', on = 'url') # Use inner to only keep records that match both datasets
print("Combined TED Summary Shape: {}".format(TED_combined.shape))

Combined TED Summary Shape: (2467, 18)


We now have 2,467 combined rows and 18 columns in the combined dataset.

In [102]:
# Examine the combined dataset for missing values and column data types
TED_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2467 entries, 0 to 2466
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   comments            2467 non-null   int64 
 1   description         2467 non-null   object
 2   duration            2467 non-null   int64 
 3   event               2467 non-null   object
 4   film_date           2467 non-null   int64 
 5   languages           2467 non-null   int64 
 6   main_speaker        2467 non-null   object
 7   name                2467 non-null   object
 8   num_speaker         2467 non-null   int64 
 9   published_date      2467 non-null   int64 
 10  ratings             2467 non-null   object
 11  related_talks       2467 non-null   object
 12  speaker_occupation  2461 non-null   object
 13  tags                2467 non-null   object
 14  title               2467 non-null   object
 15  url                 2467 non-null   object
 16  views               2467

There are **no null values** out of the 2,467 entries, however, **Speaker Occupation appears to be missing 6 values** which could be NAs and will require further analysis.

The variables with Object datatypes can be converted to categorical values, saving space and improving EDA presentation.

### Convert Date Colums to DateTime then Object Data Type to Category

In [103]:
# Update Category datatype, for data columns, to DateTime - initial dataset includes seconds so set unit = 's'
TED_combined['film_date'] = pd.to_datetime(TED_combined['film_date'], unit = 's')
TED_combined['published_date'] = pd.to_datetime(TED_combined['film_date'], unit = 's')

In [104]:
# Update Object datatype to Category
for columns in TED_combined:
    if TED_combined[columns].dtype == 'object':
        TED_combined[columns] = TED_combined[columns].astype('category')

In [105]:
# Reexamine DataFrame ensuring all datatypes previously Object are now Categorical
TED_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2467 entries, 0 to 2466
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   comments            2467 non-null   int64         
 1   description         2467 non-null   category      
 2   duration            2467 non-null   int64         
 3   event               2467 non-null   category      
 4   film_date           2467 non-null   datetime64[ns]
 5   languages           2467 non-null   int64         
 6   main_speaker        2467 non-null   category      
 7   name                2467 non-null   category      
 8   num_speaker         2467 non-null   int64         
 9   published_date      2467 non-null   datetime64[ns]
 10  ratings             2467 non-null   category      
 11  related_talks       2467 non-null   category      
 12  speaker_occupation  2461 non-null   category      
 13  tags                2467 non-null   category    

- All columns correctly reflect **Category if previously an Object datatype, excluding Film Date and Published Date** which are set to datetime datatypes.

In [106]:
# Inspect total count of null values in dataset
TED_combined.isnull().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    6
tags                  0
title                 0
url                   0
views                 0
transcript            0
dtype: int64

In [107]:
# Inspect total count of NA (NaN) values in dataset
TED_combined.isna().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    6
tags                  0
title                 0
url                   0
views                 0
transcript            0
dtype: int64

Due to the fact that there are **only 6 speakers with missing occupations**, all of whom can likely be searched online for generic occupations, these entries can be **manually encoded**.

In [108]:
TED_combined['main_speaker'][TED_combined['speaker_occupation'].isna() == True].values.tolist()

['Sonaar Luthra',
 'Rick Falkvinge',
 'Gary Kovacs',
 'Ryan Holladay',
 'Brian Dettmer',
 'Boniface Mwangi']

In [109]:
# Create a list of speakers with missing occupations
speakers = TED_combined['main_speaker'][TED_combined['speaker_occupation'].isna() == True].values.tolist()

# create an, ordered, input list for respective speaker occupations (from online search)
occupations_input = ['Designer', 'Entrepreneur', 'CEO', 'Inventor', 'Artist', 'Journalist']

# for loop to update each speaker with respective occupation that was missig
for speaker in speakers:
    for occupation in occupations_input:
        TED_combined['speaker_occupation'][TED_combined['main_speaker'] == speaker] = occupation

In [110]:
# Recheck the count of NAs to ensure above loop removed all
TED_combined.isna().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    0
tags                  0
title                 0
url                   0
views                 0
transcript            0
dtype: int64

There are no missing values in the Speaker Occupation column.

In [111]:
# Inspect total unique counts by column

TED_combined.nunique()

comments               555
description           2464
duration              1059
event                  320
film_date              690
languages               65
main_speaker          2088
name                  2464
num_speaker              5
published_date         690
ratings               2464
related_talks         2464
speaker_occupation    1421
tags                  2447
title                 2464
url                   2464
views                 2464
transcript            2464
dtype: int64

In [112]:
# Inspect Number of Speakers (per talk) column (only column with low unique values count)
TED_combined['num_speaker'].value_counts()

1    2412
2      46
3       5
4       3
5       1
Name: num_speaker, dtype: int64

**Observations**
- 320 unique events listed (e.g. TED2017)
- There 65 unique languages and 555 unique comments tracked indicate **unique counts of languages/comments per event (e.g. 4 different languages and 10 comments for a given TED talk)**
- As expected, the majority of speakers per talk is 1 (2,412), however there were 46 cases of 2 speakers presenting

In [113]:
# Run descriptive statistics summary for all variables - including categorical
TED_combined.describe(include = 'all').T

count unique  \
comments            2467    NaN   
description         2467   2464   
duration            2467    NaN   
event               2467    320   
film_date           2467    690   
languages           2467    NaN   
main_speaker        2467   2088   
name                2467   2464   
num_speaker         2467    NaN   
published_date      2467    690   
ratings             2467   2464   
related_talks       2467   2464   
speaker_occupation  2467   1421   
tags                2467   2447   
title               2467   2464   
url                 2467   2464   
views               2467    NaN   
transcript          2467   2464   

                                                                  top freq  \
comments                                                          NaN  NaN   
description         Shame is an unspoken epidemic, the secret behi...    2   
duration                                                          NaN  NaN   
event                                                         TED2014   84   
film_date                                         2017-04-24 00:00:00   64   
languages                                                         NaN  NaN   
main_speaker                                             Hans Rosling    9   
name                Jonathan Haidt: Religion, evolution, and the e...    2   
num_speaker                                                       NaN  NaN   
published_date                                    2017-04-24 00:00:00   64   
ratings             [{'id': 9, 'name': 'Ingenious', 'count': 577},...    2   
related_talks       [{'id': 1327, 'hero': 'https://pe.tedcdn.com/i...    2   
speaker_occupation                                             Writer   45   
tags                ['entertainment', 'live music', 'music', 'perf...    3   
title                                              Listening to shame    2   
url                 https://www.ted.com/talks/rob_reid_the_8_billi...    2   
views                                                             NaN  NaN   
transcript          I'm going to tell you a little bit about my TE...    2   

                        first       last        mean          std     min  \
comments                  NaT        NaT     192.571       284.83       2   
description               NaT        NaT         NaN          NaN     NaN   
duration                  NaT        NaT      821.76       334.09     135   
event                     NaT        NaT         NaN          NaN     NaN   
film_date          1984-02-02 2017-08-27         NaN          NaN     NaN   
languages                 NaT        NaT     28.2914      8.22976       1   
main_speaker              NaT        NaT         NaN          NaN     NaN   
name                      NaT        NaT         NaN          NaN     NaN   
num_speaker               NaT        NaT     1.02797     0.208371       1   
published_date     1984-02-02 2017-08-27         NaN          NaN     NaN   
ratings                   NaT        NaT         NaN          NaN     NaN   
related_talks             NaT        NaT         NaN          NaN     NaN   
speaker_occupation        NaT        NaT         NaN          NaN     NaN   
tags                      NaT        NaT         NaN          NaN     NaN   
title                     NaT        NaT         NaN          NaN     NaN   
url                       NaT        NaT         NaN          NaN     NaN   
views                     NaT        NaT  1.7403e+06  2.52709e+06  155895   
transcript                NaT        NaT         NaN          NaN     NaN   

                       25%          50%          75%          max  
comments                63          117        222.5         6404  
description            NaN          NaN          NaN          NaN  
duration               582          852         1047         3608  
event                  NaN          NaN          NaN          NaN  
film_date              NaN          NaN          NaN          NaN 

**Initial Sample Observations**
- **Hans Rosling** was the most frequent speaker (9 presentations)
- The most common speaker occupation is Writer (45 occurrences)
- Thbe most frequent date for filming, and subsequently publishing, was **4/24/2017 (64 occurrences)**
- The most frequest event (year) is **TED2014 (84 occurrences)**

### Top 5 Events, Speakers, and Occupations

In [114]:
# Inspect the top 5 speaker occupations
TED_combined['event'].value_counts().head(5)

TED2014    84
TED2009    81
TED2013    77
TED2016    75
TED2015    75
Name: event, dtype: int64

In [115]:
# Inspect the top 5 main speakers
TED_combined['main_speaker'].value_counts().head(5)

Hans Rosling      9
Juan Enriquez     7
Marco Tempest     6
Rives             6
Jonathan Haidt    5
Name: main_speaker, dtype: int64

In [116]:
# Inspect the top 5 speaker occupations
TED_combined['speaker_occupation'].value_counts().head(5)

Writer          45
Journalist      38
Artist          34
Designer        34
Entrepreneur    30
Name: speaker_occupation, dtype: int64

### Top 10 Talks of All Time

In [117]:
# Summary of top 10 viewed TED talks
TED_combined[['main_speaker', 'title', 'views']].sort_values('views', ascending = False).head(10)

main_speaker                                              title  \
0          Ken Robinson                        Do schools kill creativity?   
1268          Amy Cuddy           Your body language may shape who you are   
649         Simon Sinek                   How great leaders inspire action   
800         Brené Brown                         The power of vulnerability   
444          Mary Roach             10 things you didn't know about orgasm   
1695    Julian Treasure         How to speak so that people want to listen   
198   Jill Bolte Taylor                               My stroke of insight   
5          Tony Robbins                               Why we do what we do   
2033       James Veitch  This is what happens when you reply to spam email   
1338    Cameron Russell  Looks aren't everything. Believe me, I'm a model.   

         views  
0     47227110  
1268  43155405  
649   34309432  
800   31168150  
444   22270883  
1695  21594632  
198   21190883  
5     20685401  
2033  20475972  
1338  19787465

### Reducing Dataset - Remove Unneccessary Rows/Columns
- Since almost 98% of the Talks involved only 1 speaker, all talks with a higher count can be removed
- The following columns **don't benefit the purposes of this project as it relates to Natural Language Processing:**
    - Related Talks
    - URL
    - Languages
    - Number of Speakers    
    - Views - although helpful in identifying top videos above), offers no NLP analysis value
    - Speaker Name (redundant - already in Main Speaker column)

In [118]:
# Drop the above listed columns from the combined DataFrame
TED_combined.drop(columns = ['related_talks', 'languages', 'url','name','views', 'num_speaker'], axis = 1, inplace = True)

In [119]:
# Inspect updated DataFrame shape
TED_combined.shape

(2467, 12)

After removing the 6 columns mentioned above, the DataFrame now consists of **2.467 rows and 12 columns**.

In [120]:
# Create empty count lists and loop through appending textstat counts accordingly
sentence_cnt = []
character_cnt = []
word_cnt = []

for i in TED_combined['transcript']:
    
    sentence_cnt.append(textstat.sentence_count(i))
    character_cnt.append(textstat.char_count(i, ignore_spaces = True))
    word_cnt.append(len(i.split()))

In [121]:
# Add counts to main DataFrame
TED_combined['sentence_cnt'] = sentence_cnt
TED_combined['character_cnt'] = character_cnt
TED_combined['word_cnt'] = word_cnt

In [122]:
# Examine the top 5 rows to confirm the 3 new columns above correctly mapped
TED_combined.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event  film_date   main_speaker published_date  \
0  TED2006 2006-02-25   Ken Robinson     2006-02-25   
1  TED2006 2006-02-25        Al Gore     2006-02-25   
2  TED2006 2006-02-24    David Pogue     2006-02-24   
3  TED2006 2006-02-26  Majora Carter     2006-02-26   
4  TED2006 2006-02-22   Hans Rosling     2006-02-22   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                          transcript  sentence_cnt  \
0  Good morning. How are you?(Laughter)It's been ...           191   
1  Thank you so much, Chris. And it's truly a gre...           108   
2  (Music: "The Sound of Silence," Simon & Garfun...           223   
3  If you're here today — and I'm very happy that...           152   
4  About 10 years ago, I took on the task to teac...           209   

   character_cnt  word_cnt  
0          14344      3066  
1           9726      2089  
2          15057      3253  
3          15235      3015  
4          14245      3121

In [123]:
# Convert Ratings column to individual dictionary lists of strings, by row
TED_combined['ratings'] = [ast.literal_eval(x) for x in TED_combined['ratings']]

In [124]:
# Convert tags column to individual lists of strings, by row
TED_combined['tags'] = [ast.literal_eval(x) for x in TED_combined['tags']]

In [125]:
# USE AST library to retrievbe list inside the string dtype list (ratings)
# TED_combined['ratings'] = TED_combined['ratings'].apply(lambda x: ast.literal_eval(str(x)))

In [126]:
# USE AST library to retrievbe list inside the string dtype list (ratings)
# TED_combined["tags"] = TED_combined["tags"].apply(lambda x: ast.literal_eval(str(x)))

In [127]:
# Examine the list of all Ratings assigned to the first TED talk on the 1st row
TED_combined['ratings'][0]

[{'id': 7, 'name': 'Funny', 'count': 19645},
 {'id': 1, 'name': 'Beautiful', 'count': 4573},
 {'id': 9, 'name': 'Ingenious', 'count': 6073},
 {'id': 3, 'name': 'Courageous', 'count': 3253},
 {'id': 11, 'name': 'Longwinded', 'count': 387},
 {'id': 2, 'name': 'Confusing', 'count': 242},
 {'id': 8, 'name': 'Informative', 'count': 7346},
 {'id': 22, 'name': 'Fascinating', 'count': 10581},
 {'id': 21, 'name': 'Unconvincing', 'count': 300},
 {'id': 24, 'name': 'Persuasive', 'count': 10704},
 {'id': 23, 'name': 'Jaw-dropping', 'count': 4439},
 {'id': 25, 'name': 'OK', 'count': 1174},
 {'id': 26, 'name': 'Obnoxious', 'count': 209},
 {'id': 10, 'name': 'Inspiring', 'count': 24924}]

In [128]:
# Examine the list of all Tags assigned to the first TED talk on the 1st row
TED_combined['tags'][0]

['children',
 'creativity',
 'culture',
 'dance',
 'education',
 'parenting',
 'teaching']

The Ratings column now reflects a **list of dictionary entries for each rating and respective count assigned** - e.g. Funny at 19,645 for the 1st row.

In [129]:
# Function to extract total number from ratings column (list)
def get_total(x):
    """
    Function counts and aggregates the embedded rating counts (inside a list of dictionaries) to derive the total for each row
    (TED talk).
    
    x: input column - Ratings column to iterate through

    """
    count = 0
    for i in x:
        # i = dictionary counts
        count += i['count'] # Increase by each row's dictionary 'count' key, summing up the totals
    return count

In [130]:
# Apply the above function to create a total of all Ted Ratings, by row/talk
TED_combined['total_ratings'] = TED_combined['ratings'].apply(lambda x: get_total(x))

In [131]:
TED_combined['total_ratings']

0       93850
1        2936
2        2824
3        3728
4       25620
        ...  
2462      192
2463      151
2464      136
2465      583
2466      142
Name: total_ratings, Length: 2467, dtype: int64

In [132]:
# Sort the dataset by highest count of reviews
TED_combined.sort_values(by = 'total_ratings', ascending = False).head()

comments                                        description  duration  \
0         4553  Sir Ken Robinson makes an entertaining and pro...      1164   
198       2877  Jill Bolte Taylor got a research opportunity f...      1099   
1268      2290  Body language affects how others see us, but i...      1262   
800       1927  Brené Brown studies human connection -- our ab...      1219   
649       1930  Simon Sinek has a simple but powerful model fo...      1084   

                 event  film_date       main_speaker published_date  \
0              TED2006 2006-02-25       Ken Robinson     2006-02-25   
198            TED2008 2008-02-27  Jill Bolte Taylor     2008-02-27   
1268    TEDGlobal 2012 2012-06-26          Amy Cuddy     2012-06-26   
800        TEDxHouston 2010-06-06        Brené Brown     2010-06-06   
649   TEDxPuget Sound  2009-09-17        Simon Sinek     2009-09-17   

                                                ratings  \
0     [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
198   [{'id': 22, 'name': 'Fascinating', 'count': 14...   
1268  [{'id': 23, 'name': 'Jaw-dropping', 'count': 3...   
800   [{'id': 10, 'name': 'Inspiring', 'count': 2144...   
649   [{'id': 21, 'name': 'Unconvincing', 'count': 9...   

            speaker_occupation  \
0              Author/educator   
198             Neuroanatomist   
1268       Social psychologist   
800   Vulnerability researcher   
649          Leadership expert   

                                                   tags  \
0     [children, creativity, culture, dance, educati...   
198   [biology, brain, consciousness, global issues,...   
1268  [body language, brain, business, psychology, s...   
800   [TEDx, communication, culture, depression, fea...   
649   [TEDx, business, entrepreneur, leadership, suc...   

                                         title  \
0                  Do schools kill creativity?   
198                       My stroke of insight   
1268  Your body language may shape who you are   
800                 The power of vulnerability   
649           How great leaders inspire action   

                                             transcript  sentence_cnt  \
0     Good morning. How are you?(Laughter)It's been ...           191   
198   I grew up to study the brain because I have a ...           140   
1268  So I want to start by offering you a free no-t...           227   
800   So, I'll start with this: a couple years ago, ...           170   
649   How do you explain when things don't go as we ...           187   

      character_cnt  word_cnt  total_ratings  
0             14344      3066          93850  
198           11896      2660          70665  
1268          16409      3602          65968  
800           13781      3046          60078  
649           13758      2986          56779

In [133]:
TED_combined['total_ratings'].describe()

count     2467.000000
mean      2451.769761
std       4230.185265
min         68.000000
25%        881.500000
50%       1463.000000
75%       2521.000000
max      93850.000000
Name: total_ratings, dtype: float64

In [134]:
# Identify the Video and Speaker with the lowest total rating counts
TED_combined[['title', 'main_speaker','total_ratings']][TED_combined['total_ratings'] == TED_combined['total_ratings'].min()]

title       main_speaker  \
2445  How your pictures can help reclaim lost history  Chance Coughenour   

      total_ratings  
2445             68

In [135]:
# Identify the Video and Speaker with the highest total rating counts
TED_combined[['title', 'main_speaker','total_ratings']][TED_combined['total_ratings'] == TED_combined['total_ratings'].max()]

title  main_speaker  total_ratings
0  Do schools kill creativity?  Ken Robinson          93850

#### Observations
- The most rated TED talk of all time is **Do schools kill creativity? by Ken Robinson (93,850 total reviews)**
- The least rated TED talk of all time is **How your pictures can help reclaimn lost history by Chance Coughenour (68 total reviews)**

### Isolate Unique Ratings (Inside Row List/Dictionary)

In [136]:
# Create an empty list to append all ratings from the combined dictionary, looping through each iteration in prior lists
ratings = []
for h in range(0,len(TED_combined)):
    for i in TED_combined['ratings'][h]:
        ratings.append(i['name'])

In [137]:
# Confirm that the first 28 records (first 2 rows) in the ratings list against the first two rows in the DataFrame for ratings
ratings[0:28]

['Funny',
 'Beautiful',
 'Ingenious',
 'Courageous',
 'Longwinded',
 'Confusing',
 'Informative',
 'Fascinating',
 'Unconvincing',
 'Persuasive',
 'Jaw-dropping',
 'OK',
 'Obnoxious',
 'Inspiring',
 'Funny',
 'Courageous',
 'Confusing',
 'Beautiful',
 'Unconvincing',
 'Longwinded',
 'Informative',
 'Inspiring',
 'Fascinating',
 'Ingenious',
 'Persuasive',
 'Jaw-dropping',
 'Obnoxious',
 'OK']

In [138]:
# The list below for the Name keys should match the list above, for each of the 28 records (first 2 rows)
TED_combined['ratings'][0] + TED_combined['ratings'][1]

[{'id': 7, 'name': 'Funny', 'count': 19645},
 {'id': 1, 'name': 'Beautiful', 'count': 4573},
 {'id': 9, 'name': 'Ingenious', 'count': 6073},
 {'id': 3, 'name': 'Courageous', 'count': 3253},
 {'id': 11, 'name': 'Longwinded', 'count': 387},
 {'id': 2, 'name': 'Confusing', 'count': 242},
 {'id': 8, 'name': 'Informative', 'count': 7346},
 {'id': 22, 'name': 'Fascinating', 'count': 10581},
 {'id': 21, 'name': 'Unconvincing', 'count': 300},
 {'id': 24, 'name': 'Persuasive', 'count': 10704},
 {'id': 23, 'name': 'Jaw-dropping', 'count': 4439},
 {'id': 25, 'name': 'OK', 'count': 1174},
 {'id': 26, 'name': 'Obnoxious', 'count': 209},
 {'id': 10, 'name': 'Inspiring', 'count': 24924},
 {'id': 7, 'name': 'Funny', 'count': 544},
 {'id': 3, 'name': 'Courageous', 'count': 139},
 {'id': 2, 'name': 'Confusing', 'count': 62},
 {'id': 1, 'name': 'Beautiful', 'count': 58},
 {'id': 21, 'name': 'Unconvincing', 'count': 258},
 {'id': 11, 'name': 'Longwinded', 'count': 113},
 {'id': 8, 'name': 'Informative', '

In [139]:
# Create a list of only unique ratings within the entire ratings list 
unq_ratings = [i for n, i in enumerate(ratings) if i not in ratings[:n]]

# Convert all items in the list to lower case, to match the rest of the DataFrame that they'll be added to
for rating in range(len(unq_ratings)):
    unq_ratings[rating] = unq_ratings[rating].lower()
    
# Examine the list
unq_ratings

['funny',
 'beautiful',
 'ingenious',
 'courageous',
 'longwinded',
 'confusing',
 'informative',
 'fascinating',
 'unconvincing',
 'persuasive',
 'jaw-dropping',
 'ok',
 'obnoxious',
 'inspiring']

In [140]:
print("Length of Unique Ratings List: {}".format(len(unq_ratings)))

Length of Unique Ratings List: 14


### Create Columns for Each Rating Count

In [141]:
# As shown above, there are 14 rating categories in a consistent order - we can apply lambda functions for each unique rating

TED_combined['funny'] = TED_combined['ratings'].apply(lambda x: x[0]['count'])
TED_combined['beautiful'] = TED_combined['ratings'].apply(lambda x: x[1]['count'])
TED_combined['ingenious'] = TED_combined['ratings'].apply(lambda x: x[2]['count'])
TED_combined['courageous'] = TED_combined['ratings'].apply(lambda x: x[3]['count'])
TED_combined['longwinded'] = TED_combined['ratings'].apply(lambda x: x[4]['count'])
TED_combined['confusing'] = TED_combined['ratings'].apply(lambda x: x[5]['count'])
TED_combined['informative'] = TED_combined['ratings'].apply(lambda x: x[6]['count'])
TED_combined['fascinating'] = TED_combined['ratings'].apply(lambda x: x[7]['count'])
TED_combined['unconvincing'] = TED_combined['ratings'].apply(lambda x: x[8]['count'])
TED_combined['persuasive'] = TED_combined['ratings'].apply(lambda x: x[9]['count'])
TED_combined['jaw-dropping'] = TED_combined['ratings'].apply(lambda x: x[10]['count'])
TED_combined['ok'] = TED_combined['ratings'].apply(lambda x: x[11]['count'])
TED_combined['obnoxious'] = TED_combined['ratings'].apply(lambda x: x[12]['count'])
TED_combined['inspiring'] = TED_combined['ratings'].apply(lambda x: x[13]['count'])

In [142]:
# Examine the top 5 rows of data to ensure all new ratings columns and respective counts have been added
TED_combined.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event  film_date   main_speaker published_date  \
0  TED2006 2006-02-25   Ken Robinson     2006-02-25   
1  TED2006 2006-02-25        Al Gore     2006-02-25   
2  TED2006 2006-02-24    David Pogue     2006-02-24   
3  TED2006 2006-02-26  Majora Carter     2006-02-26   
4  TED2006 2006-02-22   Hans Rosling     2006-02-22   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  [children, creativity, culture, dance, educati...   
1  [alternative energy, cars, climate change, cul...   
2  [computers, entertainment, interface design, m...   
3  [MacArthur grant, activism, business, cities, ...   
4  [Africa, Asia, Google, demo, economics, global...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                          transcript  sentence_cnt  \
0  Good morning. How are you?(Laughter)It's been ...           191   
1  Thank you so much, Chris. And it's truly a gre...           108   
2  (Music: "The Sound of Silence," Simon & Garfun...           223   
3  If you're here today — and I'm very happy that...           152   
4  About 10 years ago, I took on the task to teac...           209   

   character_cnt  word_cnt  total_ratings  funny  beautiful  ingenious  \
0          14344      3066          93850  19645       4573       6073   
1           9726      2089           2936    544        139         62   
2          15057      3253           2824    964         45        183   
3          15235      3015           3728    760        291         32   
4          14245      3121          25620   3202       1390        942   

   courageous  longwinded  confusing  informative  fascinating  unconvincing  \
0        3253         387        242         7346        10581           300   
1          58         258        113          443          413           132   
2          60         104         78          395          230           166   
3          59         105         36           53          380          1070   
4         318         110         72           67         5433          2893   

   persuasive  jaw-dropping    ok  obnoxious  inspiring  
0       10704          4439  1174        209      24924  
1          56           268   116        131        203  
2          27           146   230         54        142  
3         132           460   230         35         85  
4        4606          3736  2542        248         61

### Convert Ratings Columns to % of Total Count

In [143]:
# Convert each of the ratings columns to a % of the overall count
for rating in unq_ratings:
    TED_combined[rating] = round((TED_combined[rating] / TED_combined['total_ratings']),4)

In [144]:
# Reexamine the ratings columns to make sure that all were converted to decimals (4) representing % of total count
TED_combined.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event  film_date   main_speaker published_date  \
0  TED2006 2006-02-25   Ken Robinson     2006-02-25   
1  TED2006 2006-02-25        Al Gore     2006-02-25   
2  TED2006 2006-02-24    David Pogue     2006-02-24   
3  TED2006 2006-02-26  Majora Carter     2006-02-26   
4  TED2006 2006-02-22   Hans Rosling     2006-02-22   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  [children, creativity, culture, dance, educati...   
1  [alternative energy, cars, climate change, cul...   
2  [computers, entertainment, interface design, m...   
3  [MacArthur grant, activism, business, cities, ...   
4  [Africa, Asia, Google, demo, economics, global...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                          transcript  sentence_cnt  \
0  Good morning. How are you?(Laughter)It's been ...           191   
1  Thank you so much, Chris. And it's truly a gre...           108   
2  (Music: "The Sound of Silence," Simon & Garfun...           223   
3  If you're here today — and I'm very happy that...           152   
4  About 10 years ago, I took on the task to teac...           209   

   character_cnt  word_cnt  total_ratings   funny  beautiful  ingenious  \
0          14344      3066          93850  0.2093     0.0487     0.0647   
1           9726      2089           2936  0.1853     0.0473     0.0211   
2          15057      3253           2824  0.3414     0.0159     0.0648   
3          15235      3015           3728  0.2039     0.0781     0.0086   
4          14245      3121          25620  0.1250     0.0543     0.0368   

   courageous  longwinded  confusing  informative  fascinating  unconvincing  \
0      0.0347      0.0041     0.0026       0.0783       0.1127        0.0032   
1      0.0198      0.0879     0.0385       0.1509       0.1407        0.0450   
2      0.0212      0.0368     0.0276       0.1399       0.0814        0.0588   
3      0.0158      0.0282     0.0097       0.0142       0.1019        0.2870   
4      0.0124      0.0043     0.0028       0.0026       0.2121        0.1129   

   persuasive  jaw-dropping      ok  obnoxious  inspiring  
0      0.1141        0.0473  0.0125     0.0022     0.2656  
1      0.0191        0.0913  0.0395     0.0446     0.0691  
2      0.0096        0.0517  0.0814     0.0191     0.0503  
3      0.0354        0.1234  0.0617     0.0094     0.0228  
4      0.1798        0.1458  0.0992     0.0097     0.0024

In [145]:
# Print column summary for adding together all % columns to create total column and sum to to 100%
for i in TED_combined.columns:
    print(TED_combined.columns.get_loc(i), i)

0 comments
1 description
2 duration
3 event
4 film_date
5 main_speaker
6 published_date
7 ratings
8 speaker_occupation
9 tags
10 title
11 transcript
12 sentence_cnt
13 character_cnt
14 word_cnt
15 total_ratings
16 funny
17 beautiful
18 ingenious
19 courageous
20 longwinded
21 confusing
22 informative
23 fascinating
24 unconvincing
25 persuasive
26 jaw-dropping
27 ok
28 obnoxious
29 inspiring


Columns: 16 through 29 should be summed up and new column **total ratings (%)** created.

In [146]:
# Create new Total Ratings (%) column based on aggregating each rating column % split of total
TED_combined['total ratings (%)'] = round(TED_combined.iloc[:,16:30].sum(axis = 1),1)

In [147]:
TED_combined.head(1)

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   

     event  film_date  main_speaker published_date  \
0  TED2006 2006-02-25  Ken Robinson     2006-02-25   

                                             ratings speaker_occupation  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...    Author/educator   

                                                tags  \
0  [children, creativity, culture, dance, educati...   

                         title  \
0  Do schools kill creativity?   

                                          transcript  sentence_cnt  \
0  Good morning. How are you?(Laughter)It's been ...           191   

   character_cnt  word_cnt  total_ratings   funny  beautiful  ingenious  \
0          14344      3066          93850  0.2093     0.0487     0.0647   

   courageous  longwinded  confusing  informative  fascinating  unconvincing  \
0      0.0347      0.0041     0.0026       0.0783       0.1127        0.0032   

   persuasive  jaw-dropping      ok  obnoxious  inspiring  total ratings (%)  
0      0.1141        0.0473  0.0125     0.0022     0.2656                1.0

### Move Total Ratings column to end of DataFrame

In [148]:
# To move the column to the end, first create a copy column/series and delete the original column from the DataFrame
last_column = TED_combined.pop('total_ratings')

# Then insert the copied column at position 30 (+1 to the 29th column previously last)
TED_combined.insert(30, 'total_ratings', last_column)

In [149]:
# Examine the top row of data, ensuring that the total ratings % is 1.0 and the total ratings equals 93,850 seen previously
TED_combined.head(1)

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   

     event  film_date  main_speaker published_date  \
0  TED2006 2006-02-25  Ken Robinson     2006-02-25   

                                             ratings speaker_occupation  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...    Author/educator   

                                                tags  \
0  [children, creativity, culture, dance, educati...   

                         title  \
0  Do schools kill creativity?   

                                          transcript  sentence_cnt  \
0  Good morning. How are you?(Laughter)It's been ...           191   

   character_cnt  word_cnt   funny  beautiful  ingenious  courageous  \
0          14344      3066  0.2093     0.0487     0.0647      0.0347   

   longwinded  confusing  informative  fascinating  unconvincing  persuasive  \
0      0.0041     0.0026       0.0783       0.1127        0.0032      0.1141   

   jaw-dropping      ok  obnoxious  inspiring  total ratings (%)  \
0        0.0473  0.0125     0.0022     0.2656                1.0   

   total_ratings  
0          93850

In [159]:
unq_ratings

['funny',
 'beautiful',
 'ingenious',
 'courageous',
 'longwinded',
 'confusing',
 'informative',
 'fascinating',
 'unconvincing',
 'persuasive',
 'jaw-dropping',
 'ok',
 'obnoxious',
 'inspiring']

In [186]:
top_ratings = pd.DataFrame(columns = ['rating'], data = TED_combined[['funny', 'beautiful', 'ingenious', 'courageous', 'longwinded', 'confusing', 'informative', 'fascinating',
 'unconvincing', 'persuasive', 'jaw-dropping', 'ok', 'obnoxious', 'inspiring']].max())

top_ratings

rating
funny         0.7021
beautiful     0.6699
ingenious     0.8332
courageous    0.4244
longwinded    0.4809
confusing     0.4145
informative   0.3945
fascinating   0.4542
unconvincing  0.4268
persuasive    0.6428
jaw-dropping  0.7100
ok            0.1808
obnoxious     0.4504
inspiring     0.3864

In [200]:
TED_combined[['title', 'main_speaker', 'event', 'rating']][TED_combined['funny'] == 0.7021]
TED_combined[['title', 'main_speaker', 'event', 'rating']][TED_combined['beautiful'] == 0.6699]

KeyError: "['rating'] not in index"

In [183]:
TED_combined[['title', 'main_speaker', 'event']](
    [TED_combined['funny'] == TED_combined['funny'].max()] or
    [TED_combined['beautiful'] == TED_combined['beautiful'].max()])

TypeError: 'DataFrame' object is not callable

In [175]:
events_by_top_ratings = pd.merge(TED_combined, top_ratings, how = 'inner', \
                                 on = [['funny', 'beautiful', 'ingenious', 'courageous', 'longwinded', 'confusing', 'informative', 'fascinating' \
                                        , 'unconvincing', 'persuasive', 'jaw-dropping', 'ok', 'obnoxious', 'inspiring']].max()) # Use inner to only keep records that match both datasets
print("Combined TED Summary Shape: {}".format(TED_combined.shape))

AttributeError: 'list' object has no attribute 'max'

In [172]:
top_rating_df = pd.DataFrame(data = TED_combined[['title', 'main_speaker', 'event']], index = [])
top_rating_df

Empty DataFrame
Columns: [title, main_speaker, event]
Index: []

In [157]:
TED_combined[['title', 'main_speaker', 'event', 'funny']][TED_combined['funny'] == TED_combined['funny'].max()],
TED_combined[['title', 'main_speaker', 'event', 'beautiful']][TED_combined['funny'] == TED_combined['beautiful'].max()]

(                        title   main_speaker    event   funny
 657  It's time for "The Talk"  Julia Sweeney  TED2010  0.7021,
 Empty DataFrame
 Columns: [title, main_speaker, event, beautiful]
 Index: [])

In [153]:
top_rating_df = pd.DataFrame(TED_combined[['title', 'main_speaker', 'event', 'rating']])

for rating in unq_ratings:
    top_rating_df = TED_combined[['title', 'main_speaker', 'event', rating]] \
                        [TED_combined[rating] == TED_combined[rating].max()]
    
top_rating_df   

KeyError: "['rating'] not in index"

In [ ]:
sent = sent_tokenize(TED_combined['transcript'][0])
sent

In [ ]:
# Tokenize Transcripts and Remove Stopwords
stop_words = set(stopwords.words('english'))

result = [i for i in sent if not i in stop_words]
print(result)

In [ ]:
# Tokenize Transcripts and Remove Stopwords

stop_words = set(stopwords.words('english'))
tokens = word_tokenize(TED_combined['transcript'][0])
result = [i for i in tokens if not i in stop_words]
print(result)

In [ ]:
for line in result:
    re.sub(pattern = '\W', repl = '', string = line)
    print(line)

In [ ]:
input_str = 'NLTK is a leading platform for building Python programs to work with human language data.'
stop_words = set(stopwords.words('english'))
from nltk.tokenize import word_tokenize
tokens = word_tokenize(input_str)
result = [i for i in tokens if not i in stop_words]
print (result)

In [ ]:
stemmer= PorterStemmer()
for word in result:
    print(stemmer.stem(word))

In [ ]:
TED_combined['transcript'][0].lower().split()